### DISCUSSION TITLE
Swapping hotkeys not working when mouse button held down

### DISCUSSION THREAD

**Discussion by dlogneb** [pqrs-org/Karabiner-Elements#3581] [Q&A]
I'm using Karabiner to swap the command and control keys on my Mac when remotely accessing a PC with HP RGS.   It works great except for one use case:

In the app I'm working on when you click and drag an object you can press control to make it "snap" to other objects.
But once I've swapped the keys with Karabiner I lose this functionality.  If I click and drag and then press the command key (which is control) it does not work.  If I hold down the command(control) key _first_ and then click and drag it does work. 

Once the mouse button is pressed, the additional key doesn't work.  With Karabiner disabled, the key press (actual control key) _does_ work.   Confusingly, with Karabiner enabled if I hold Shift-Command down the application _does_ behave how it is supposed to when holding Shift-Control (it locks the segment on one axis).   So I'm pretty puzzled!

Any thoughts?

Here is my json file:

{
  "title": "Swap Left Control/Command only in HP RGS",
  "rules": [
    {
      "description": "Left Command to Control",
      "manipulators": [
        {
          "type": "basic",
          "from": {
            "key_code": "left_command"
          },
          "to": [
            {
              "key_code": "left_control"
            }
          ],
          "conditions": [
            {
              "type": "frontmost_application_if",
              "bundle_identifiers": [
                "^com.hp.rgs.HP-RGS-Receiver"
              ]
            }
          ]
        }
      ]
    },
    {
      "description": "Left Control to Command",
      "manipulators": [
        {
          "type": "basic",
          "from": {
            "key_code": "left_control"
          },
          "to": [
            {
              "key_code": "left_command"
            }
          ],
          "conditions": [
            {
              "type": "frontmost_application_if",
              "bundle_identifiers": [
                      "^com.hp.rgs.HP-RGS-Receiver"
              ]
            }
          ]
        }
      ]
    }
  ]
}

**Comment by rawb24** [comment]
Hi! I'm sorry I don't have an answer for your issue but I was wondering if you could actually help me? 

I just became familiar with Karabiner. Starting Monday I need to remote into a PC computer via HP RGS. I'm on Mac. I couldn't find anything about swapping the control/command keys (only while using RGS). 

I don't know anything about coding, or even GitHub. Could you tell me how to install this function in Karabiner, please? I figure I could copy/add your .json to Karabiner but I have no idea how to do that, or if that would even work. Thanks so much.

**Reply by dlogneb** [reply]
Sure.  Just know that I am no expert. 

I've attached my json file (zipped) in this message.   Uncompress it (so it ends with .json) and place it in:

.config/karabiner/assets/complex_modifications/

To get to the .config folder open your terminal app and type "cd .config" press return and then "open ." and press return.

Once you've placed the .json file, go to Karabiner settings and choose Complex Modifications.  Press the "add rule" button and then you should see the two rules I have.   Like in this picture:  https://imgur.com/a/bKfUXdU

Enable both rules.

Good luck!

[ComConSwapHPzCentral.json.zip](https://github.com/pqrs-org/Karabiner-Elements/files/12559159/ComConSwapHPzCentral.json.zip)

**Comment by serdarsen** [comment]
Hi @dlogneb, 

While I'm not familiar with HP RGS and I'm relatively new to Karabiner, I'd like to share my experiences. 

Is your "Modify event" option for the mouse enabled as discussed in [Discussion: Remapping Control and Left Mouse Button to Command and Left Mouse Button Issue (comment)](https://github.com/pqrs-org/Karabiner-Elements/discussions/3589#discussioncomment-6897639)?

**Reply by dlogneb** [reply]
Currently it is not.   However turning on "modify event" does not rectify the issue.  Turning it on also slows my mouse down for some reason, any idea why that may be happening?

I'm not sure if you're suggesting turning on "modify events" should fix the issue, or if I also have to add json rules like the person in that discussion did.

To clarify my issue – holding the modifier key down AND THEN holding the mouse button works, holding the mouse button and then the modifier key does not work.   

thanks!

**Reply by serdarsen** [reply]
I initially thought that giving Karabiner full control of the mouse by enabling 'modify event' and including pointing_button in the JSON configuration could potentially solve the issue. However, I've had some experiences with configurations similar to yours, and I'm no longer certain about this approach. I'm not sure why enabling 'modify events' would slow down your mouse; that's indeed puzzling. I understand the issue may involve configuring the order of events, but I must admit that I don't have an in-depth understanding of Karabiner to provide a definitive solution. Hopefully, someone with more expertise can chime in to assist you further

**Comment by serdarsen** [comment]
Hi @drmmrboy24,

While I'm not familiar with HP RGS and I'm relatively new to Karabiner, I'd like to share my experiences. You can easily install Karabiner Elements using Homebrew and create your custom complex modifications.

1. Install and configure Karabiner Elements using one of the following methods:

   - **Official Documentation**: For detailed installation and configuration instructions, refer to the [official documentation: Getting Started](https://karabiner-elements.pqrs.org/docs/getting-started/).

   - **Homebrew (recommended)**: If you prefer using Homebrew, you can install Karabiner Elements with the [Homebrew Formulae: karabiner-elements](https://formulae.brew.sh/cask/karabiner-elements). Also, Alex teaching how to install Homebrew in [Youtube: Setting up new MacBook for software development (2.25)](https://youtu.be/mmkDyV59nRo?si=ubV4h1cVWYeWVsBG&t=145)
     This method simplifies the installation process and keeps your software up-to-date.

2. Create your custom complex modifications in Karabiner Elements:
   - You can watch this informative video tutorial [YouTube: How to Create Your Own CUSTOM Complex Modifications in Karabiner-elements](https://www.youtube.com/watch?v=iiSIaMD4vqY).
   - You can review my custom configurations in [serdarsen/karabiner](https://github.com/serdarsen/karabiner) repository. You can learn your key_code names using [Karabiner EventViewer](https://karabiner-elements.pqrs.org/docs/manual/operation/eventviewer/).

**Comment by rawb24** [comment]
Thank you so much for this info. I will try it out. I greatly appreciate it!


On Sep 7, 2023, at 10:05 PM, Serdar ŞEN ***@***.***> wrote:

﻿

Hi @drmmrboy24<https://github.com/drmmrboy24>

You can install Karabiner Elements using Homebrew easly and create your custom complex modifications.

  1.  Install and configure Karabiner Elements using one of the following methods:

     *   Official Documentation: For detailed installation and configuration instructions, refer to the official documentation: Getting Started<https://karabiner-elements.pqrs.org/docs/getting-started/>.

     *   Homebrew (recommended): If you prefer using Homebrew, you can install Karabiner Elements with the Homebrew Formulae: karabiner-elements<https://formulae.brew.sh/cask/karabiner-elements>. Also, Alex teaching how to install Homebrew in Youtube: Setting up new MacBook for software development (2.25)<https://youtu.be/mmkDyV59nRo?si=ubV4h1cVWYeWVsBG&t=145>
This method simplifies the installation process and keeps your software up-to-date.

  2.  Create your custom complex modifications in Karabiner Elements:

     *   You can watch this informative video tutorial YouTube: How to Create Your Own CUSTOM Complex Modifications in Karabiner-elements<https://www.youtube.com/watch?v=iiSIaMD4vqY>.
     *   You can review my custom configurations in serdarsen/karabiner<https://github.com/serdarsen/karabiner> repository. You can learn your key_code names using Karabiner EventViewer<https://karabiner-elements.pqrs.org/docs/manual/operation/eventviewer/>.

**Comment by rawb24** [comment]
Wow, I really appreciate you. Thank you so much. I hope this works with my new job. Thanks again!


On Sep 8, 2023, at 4:55 AM, dlogneb ***@***.***> wrote:

﻿

Sure. Just know that I am no expert.

I've attached my json file (zipped) in this message. Uncompress it (so it ends with .json) and place it in:

.config/karabiner/assets/complex_modifications/

To get to the .config folder open your terminal app and type "cd .config" press return and then "open ." and press return.

Once you've placed the .json file, go to Karabiner settings and choose Complex Modifications. Press the "add rule" button and then you should see the two rules I have. Like in this picture: https://imgur.com/a/bKfUXdU

Enable both rules.

Good luck!

ComConSwapHPzCentral.json.zip<https://github.com/pqrs-org/Karabiner-Elements/files/12559159/ComConSwapHPzCentral.json.zip>
